syntax/parse.go

```diff
@@ -938,3 +947,99 @@ func terminatesExprList(tok Token) bool {
 	}
 	return false
 }
+
+// Comment assignment.
+// We build two lists of all subexpressions, preorder and postorder.
+// The preorder list is ordered by start location, with outer expressions first.
+// The postorder list is ordered by end location, with outer expressions last.
+// We use the preorder list to assign each whole-line comment to the syntax
+// immediately following it, and we use the postorder list to assign each
+// end-of-line comment to the syntax immediately preceding it.
+
+// flattenAST returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAST(f *File) (pre, post []Node) {
+	stack := []Node{}
+	Walk(f, func(n Node) bool {
+		if n != nil {
+			pre = append(pre, n)
+			stack = append(stack, n)
+		} else {
+			post = append(post, stack[len(stack)-1])
+			stack = stack[:len(stack)-1]
+		}
+		return true
+	})
+	return pre, post
+}
+
+// assignComments attaches comments to nearby syntax.
+func (p *parser) assignComments(f *File) {
+	pre, post := flattenAST(f)
+
+	// Split into whole-line comments and suffix comments.
+	var line, suffix []Comment
+	for _, com := range p.in.comments {
```

Comment:
Why not have the scanner divide the comments into the two categories?
Then you can do away with the .Suffix field.

Reply:
We need to keep the .Suffix field in the AST. This is a useful information, for example if we want to print it back.

Reply:
Well... the information seems duplicated (there's a Suffix Comment slice and the suffix boolean field). I have to investigate why Buildifier does it like this.